### DISCUSSION TITLE
Proposal for loading packages in Spack

### DISCUSSION THREAD

**Discussion by haampie** [spack/spack#26266] [Ideas]
We currently have three different ways to load packages "at runtime":

1. `spack load` 
2. `spack env activate`
3. `module load`

all of which have different behaviors. The idea of this proposal is to ensure they all
have the same semantics by default, but still allow them to be configurable in ways
that make sense for certain centers (in particular w.r.t. module files).

As similar command is `spack build-env`, but this proposal is targeted towards installed
specs and how to interact with them as a user, not as a developer.

## What does "loading a package/spec" currently mean?

We're calling `spack.user_environment.environment_modifications_for_spec`, which more or
less does the following for a given spec X:

1. Add `prefix_inspections`:
   Prepends the (subdir of the) X's prefix to `PATH`, `CMAKE_PREFIX_PATH` etc as
   specified by the user in their `modules.yaml`.
2. Add `modifications_from_dependencies` in a *run*-context:
   From the docstring it "returns the environment modifications that are required by the
   dependencies of a spec and also applies modifications to this spec's package at
   module scope."
   This code is very hard to follow and more targeted towards build/test context than
   run context, my TL;DR version:
   1. For *all transitive run-type deps* of X: it adds `<prefix>/bin[64]` to the PATH.
   2. For *all transitive link-and-run-type deps* of X it calls:
      a. `set_module_variables_for_package`
         (This defines props on the dependency, like `make`, `ninja`, # build jobs, 
          cmake standard args, `spack_cc` executable, ... all related to builds.)
      b. `setup_dependent_package`
         (This makes the dependency set a bunch of props on the current package or on
          the dep itself, usually used as `self.spec.mpicc = xyz`, so the dependent can
          use that during the build)
      c. `setup_dependent_run_environment`
         (This is used to make the dependency modify env variables that may depend on
          the dependent = current package's spec. E.g. if py-x is loaded and it depends
          on Python, then the Python package iterates over all depedencies of py-x to
          find more python packages, and adds those paths to `PYTHONPATH`.)
      
      It does **not** call `setup_run_environment` on these transitive dependencies.
3. Call `setup_run_environment` of X only (non-recursively)

## What packages are loaded by default without changing config?

1. `spack load x y z` collects *all transitive dependencies of any type* of CLI specs and
   loads these packages individually (so, each individual package load recurses again as the astute reader notices :D)
2. `spack env activate env` collects *all transitive run-type dependencies of all env
   root specs* and loads these packages individually (same note as above)
3. `module load x y z` *only* loads the CLI-provided specs.

## What are the current problems with loading?

Short version:

a. Inconsistent behavior (as outlined above)
b. Risk of broken module files by default
c. Bad performance
d. `setup_dependent_run_environment` is more of a hack to patch module files

**Module files issues**
   Runtime package loading *requires* specs of transitive run-type to be loaded.
   
   Loading run-type dependencies shouldn't be interpreted weakly like putting
   `<prefix>/bin` in the path, but rather: it requires
   `Package::setup_[dependendent_]run_environment` to be called on all transitive
   run-type deps, since the package knows/defines what is necessary at runtime apart
   from the sensible defaults Spack or the user sets in `prefix_inspections`.

   Our generated module files do not contain the modifications from
   `setup_run_environment`, only `setup_dependent_run_environment`. If you have `A`
   depends on `B` depends on `C` all with run deptype, then by default,
   `module load A` loads `A` but not `B` or `C`.

   You get false impressions from `module load py-xyz`: module files for Python
   packages usually work, not because we recursively load run-type deps, but because
   the Python packages sets `PYTHONPATH` for all run-type deps in
   `setup_dependent_run_environment`. However, a correct `PYTHONPATH` is not enough, it
   is still required to call `setup_run_environment` on every dependency, and Python
   packages that specify an elaborate setup there are broken as dependencies when using
   `module load`.

   If we would make `setup_dependent_run_environment` call `setup_run_environment`, we
   effectively get working self-contained module files, but then we also run
   `setup_run_environment` twice in `spack load` / `spack env activate` for all
   packages.

   Note that Python packages can easily work without `setup_dependent_run_environment`,
   if they just set add their `PYTHONPATH` in `setup_run_environment`, which can be
   done in the base class Python packages inherit. For modules to work though, this
   implies that all run-type dependencies are auto-loaded.

**Unclear what goes into `setup_dependent_run_environment` / `setup_run_environment`**
   For the user it is in fact unclear when to use `setup_dependent_run_environment` vs
   `setup_run_environment`, and often you see packages implement both with duplicated
   contents, without the dependent spec from `setup_dependent_run_environment` being
   used.

**Performance issues**
   Loading the runtime environment should be super fast, which is the case with module
   files as they are mostly static, but not the case with `spack load`,
   `spack env activate`, and generating module files.

   `spack load` clearly loads too many packages, it makes more sense to stick to
   loading the roots + transitive run-type deps only, just like in `spack env activate`.

   `spack env activate` is slow because it lists all run-type packages, and then loads
   each of them individually, which recurses into their dependencies, leading to `n^2`
   visits of all packages if there's a chain of `n` of them. Instead, we should be able
   to visit each package once during a load.

   Every Python package in the DAG will call `setup_dependent_run_environment` on
   Python itself, which traverses the dependencies again. This is yet another redundant
   traversal of all packages. We can delegate `setup_dependent_run_environment` to
   each package's `setup_run_environment` instead. Similar issues exist for R, Perl,
   Java, ... packages -- they can be fixed on the package level itself.

   Loading does a lot of calls to EnvironmentModifications methods, which unwind the
   stracktrace to keep track of what packages modify what, which is useful in builds
   to warn the user that a package by-passes Spack's compiler wrappers. For run-time
   loading of already installed packages this makes no sense, and it's a huge
   performance killer.

   Runtime loading still sets up all packages as if they were going to be built. In
   particular:
   1. `set_module_variables_for_package`
   2. `setup_dependent_package` (I've only seen it used in a build context)

## Proposal

To make our lives easier and to make users know what to expect, we should fix the
inconsistencies between the defaults `spack load`, `spack env activate` and
`module load`.

### Recursive module loading by default
Most controversial, I would want to suggest to make `module load` autoload its run-type*
deps by default. This way, every module file contains the environment changes for its
Spack package and no more. So the module file for `py-xyz` appends its prefix to the
`PYTHONPATH` (done by `setup_run_environment`) but not those of its dependencies.

This is well supported by `lmod`, since it has `depends_on` and does proper reference
counting of loaded modules. It seems very useful and trivial to enable in Spack.

For `tcl`, autoloading support is somewhat hacky, but we have it, it's just that there's
no refcounting. Users can still disable autoloading, and instead module load everything
by hand -- but clearly with Python packages that'd be unmanageable. We *could* support
module files that basically "inline" the `depends_on`'s that lmod uses, so that you have
1 module file which contains all changes for all transitive dependencies, but this looks
somewhat questionable.

\* Note: currently it also loads link-type deps by default -- do we need that? Don't we have rpath?

To be figured out: should external (cray) modules be `depends_on`'d?

### Loading should not require any build related changes

I'm reasonably confident that `setup_dependent_package` is only used in a build context
to modify a Spec to make executables available. It'd be nice to skip it. Also we could
drop `set_module_variables_for_package`, it's very build-related.

### No more `setup_dependent_run_environment`

I think `setup_dependent_run_environment` is redundant. It's causing performance
regressions because of walking the DAG excessively in Python/R/etc packages,
and it is used to insert environment changes in the parent module file -- but this is
no long necessary when we do autoloading. Finally it's confusing users and we see many
packages just duplicating `setup_run_environment` in `setup_dependent_run_environment`,
since it's unclear in what context each is used.

### Visiting packages to be loaded once in a deterministic order

When running `spack load x y z` or `spack env activate env` with `[x, y, z]` the root
specs, and `setup_dependent_run_environment` + `setup_dependent_package` +
`set_module_variables_for_package` removed or dropped as requirements for runtime
loading, we can in fact collect the unique list of roots with transitive run-type deps
ahead of time in a sensible order (post-order with @scheibelp's pessimisation of
externals), loop over them *once*, and call `prefix_inspections` and
`setup_run_environment` once, and call it a day.

**Comment by haampie** [comment]
One point of discussion (thanks @becker33):

Some libraries may require environment variables to work. You could argue:
a) the dependent should make them (link, run)-type deps then, and if there is a chain `A` <--link-- `B` <--link,run-- `C`, then `setup_run_environment` on `C` should be called when `A` is loaded (so, we load all run-type deps reachable from link/run edges).
b) `setup_run_environment` should always be called on link-type dependencies.

If (b) then what's the difference between a run and a link-type dep?

**Comment by scheibelp** [comment]
Some use cases and considerations:

* Link dependencies may want to modify LIBRARY_PATH (which RPATH doesn’t effect)
* > `* Note: currently it also loads link-type deps by default -- do we need that? Don't we have rpath?`

  Generally yes although it is nice to have a simple mechanism available for modifying LD_LIBRARY_PATH in cases where we suspect issues with RPATH (some systems have ignored RPATH in some cases). That doesn’t mean is has to be the default behavior though
* Packages which override setup_run_environment will have to be careful to invoke their super() implementation (whereas before one wouldn’t need to consider inheritance since e.g. the Spack Python package handles this in setup_dependent_environment)
    - It’s worth noting this since package.py is designed to be as “friendly” as possible to users with limited Python experience, but this is the sort of think we could also catch in code review
    - Overall though I agree that setup_dependent_run_environment and setup_run_environment can be consolidated
- https://github.com/spack/spack/pull/14062 updated spack load to no-longer be a wrapper around module load, there are some behaviors of lmod modules which we could likely support but AFAIK we don’t currently support with `spack load`
    - For example if we `spack load zlib%gcc`, we don’t force loading `zlib%intel` based on loading `hdf5%intel`, the lmod hierarchy created by Spack is capable of that
    - That previous example is complicated by Spack’s special handling of compilers, but I think the same issue applies if you `spack load hdf5^mvapich` and then `spack load openmpi`: the lmod hierarchy can force switching of modules
    - So it’s probably easy to update env modifications so `spack env activate` works like `spack load /hash-of-root-a /hash-of-root-b…` but it might be hard to say that `spack load` has the same behavior as lmod
    - To that end, `spack env deactivate` will undo all the modifications, so in that sense behaves similar to lmod conflict resolution

Other comments:

> Most controversial, I would want to suggest to make module load autoload its run-type* deps by default. 

This seems fine to me, especially if users can change that

> For tcl, autoloading support is somewhat hacky, but we have it, it's just that there's no refcounting

IMO this isn’t a major problem: I would agree that initially-working modules is more important than supporting cleanup; we could suggest that users enter a new shell when loading spack-created TCL modules

> If (b) then what's the difference between a run and a link-type dep?

https://github.com/spack/spack/issues/3768 (and https://github.com/spack/spack/issues/3768#issuecomment-293046651) outlines some differences 

> You could argue: a) the dependent should make them (link, run)-type deps then

That should probably fall out of a resolution to 3768 (although as noted in that issue, there’s still a couple things that we want to do for link deps that we don’t want to do for run deps)

> Also we could drop set_module_variables_for_package, it's very build-related.

I think it is also useful for inspecting the package in `spack-python`. I don't think it's needed though when adding the package to the user's shell.

> [modifications_from_dependencies] is very hard to follow and more targeted towards build/test context than run context, my TL;DR version:

This could use some elaboration: I’m confused because this function takes `context` as an argument, and it does different things when the `context == ‘run’`; this means pulling in different sets of dependencies based on their type; the function is useful since there are ordering considerations which I assume apply when building and for when running the root package (e.g. making sure that a merged prefix of an external doesn’t obscure Spack-built packages from the path). 

* This might be an example of something Spack load does that is difficult to replicate with `module load` (at least without changing the auto-generated module files)

**Comment by scheibelp** [comment]
We had some discussion about this today in https://github.com/spack/spack/wiki/Telcon:-2021-10-06. Probably the easiest thing we can do for now is to enable autoloading for modules by default, but only for lmod (we can enable reference counting in lmod without much work).

We may identify other ways to more-closely align the behavior of these approaches over time, but don't need to to do it all at once.
